dali/operators/math/expressions/expression_impl_gpu.cuh

```diff
@@ -199,41 +339,205 @@ inline dim3 GetGridLayout(int extent, int tiles) {
+          sample_cpu.args[i].strides[d] = sample.args[i].strides[d];
+        }
+      };
+      prepare_args(0);
+      prepare_args(1);
+    }
+
+    std::tie(samples_gpu, tiles_gpu) = s.ToContiguousGPU(ctx.stream, samples_cpu, tiles);
+    auto grid = GetGridLayout(kBlocksX, tiles.size());
+    auto block = dim3(kThreadNum, 1, 1);
+    Invoker::Invoke(samples_gpu, tiles_gpu, grid, block, ctx.stream);
+  }
+
+ private:
+  // Use BinaryArithmeticOpGpuPerfTest for tuning
+  static constexpr int kNumArgs = 2;
+  static constexpr int kThreadNum = 256;
+  static constexpr int kBlocksX = 64;
+};
+
+template <typename Invoker>
+class ExprImplGPUInvokeTernary : public ExprImplBase {
+ public:
+  void Execute(ExprImplContext &ctx, span<const SampleDesc> samples,
+               span<const TileDesc> tiles) override {
+    int ndim = samples[0].output.shape.sample_dim();
+    VALUE_SWITCH(ndim, Dims, (1, 2, 3, 4, 5, 6), (
+      ExecuteImpl<Dims>(ctx, samples, tiles);
+    ), DALI_FAIL(make_string("Unsupported number of dimensions: ", ndim)););  // NOLINT
+  }
+
+  template <int ndim>
+  void ExecuteImpl(ExprImplContext &ctx, span<const SampleDesc> samples,
+                   span<const TileDesc> tiles) {
+    kernels::DynamicScratchpad s({}, ctx.stream);
+    assert(samples.size() > 0);
+    for (int i = 0; i < samples.size(); i++) {
+      assert(ndim == samples[i].output.shape.sample_dim());
+      assert(kNumArgs == samples[i].args.size());
+    }
+
+    auto sample_descs =
+        make_span(s.Allocate<mm::memory_kind::host, SampleDescGPU<kNumArgs, ndim>>(samples.size()),
+                  samples.size());
+    for (int i = 0; i < samples.size(); i++) {
+      auto &sample_desc = sample_descs[i];
+      auto &sample = samples[i];
+      sample_desc.output.data = sample.output.data;
+      sample_desc.output.dtype = sample.output.dtype;
+      for (int d = 0; d < ndim; d++) {
+        sample_desc.output.shape[d] = sample.output.shape[d];
+        sample_desc.output.strides[d] = sample.output.strides[d];
+      }
+
+      sample_desc.args[0].data = sample.args[0].data;
+      sample_desc.args[0].dtype = sample.args[0].dtype;
+      for (int d = 0; d < ndim; d++) {
+        sample_desc.args[0].shape[d] = sample.args[0].shape[d];
+        sample_desc.args[0].strides[d] = sample.args[0].strides[d];
+      }
+
+      sample_desc.args[1].data = sample.args[1].data;
+      sample_desc.args[1].dtype = sample.args[1].dtype;
+      for (int d = 0; d < ndim; d++) {
+        sample_desc.args[1].shape[d] = sample.args[1].shape[d];
+        sample_desc.args[1].strides[d] = sample.args[1].strides[d];
+      }
+
+      sample_desc.args[2].data = sample.args[2].data;
+      sample_desc.args[2].dtype = sample.args[2].dtype;
+      for (int d = 0; d < ndim; d++) {
+        sample_desc.args[2].shape[d] = sample.args[2].shape[d];
+        sample_desc.args[2].strides[d] = sample.args[2].strides[d];
+      }
```

Comment:
This is probably worth to put in a base class with a loop for the number of arguments, and use in all variants to reduce the boilerplate code duplication.

Something like `SampleDescToSampleDescGPU`

There is even lambda for prepare_args in 2-arg case.